RTL8710BN can't compile on ESPHome 2026.3.0

### The problem

Hi, I'm running EPSHome on a RTL8710BN on a module from Solis. 
This specific YAML to be exact: https://github.com/hn/ginlong-solis/blob/master/solis-esphome-emw3080.yaml
Just to be sure i already disabled the ringbufferfix platform_packages, but i am not able to compile with 2026.3.0.

If more info is needed i'm sorry i didn't add it but will of course when requested. 

### Which version of ESPHome has the issue?

2026.3.0

### What type of installation are you using?

Home Assistant Add-on

### What platform are you using?

RTL87XX

### Component causing the issue

_No response_

### YAML Config

```yaml
#
# solis-esphome-emw3080.yaml -- Ginlong Solis Inverter ESPhome definition for EMW3080 MCUs.
#
# The hardware described in this file (MCU and GPIO) corresponds to that of the Solis S3 WiFi stick.
#
# (C) 2023-2024 Hajo Noerenberg
#
# Usage:
#   1. Depending on your type of device (INV, ESINV or EPM),
#      include modbus definition in 'packages' section below 
#   2. Compile and upload
#      https://docs.libretiny.eu/docs/projects/esphome/#build-upload
#
#
# http://www.noerenberg.de/
# https://github.com/hn/ginlong-solis
#
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3.0 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/gpl-3.0.txt>.
#

esphome:
  name: solis-emw3080
  friendly_name: Ginlong Solis Inverter
  platformio_options:
    board_build.mcu: rtl8710bn
    board_build.f_cpu: 125000000L
    platform_packages:
      - framework-arduino-api @ https://github.com/hn/ArduinoCore-API#RingBufferFix

rtl87xx:
  board: generic-rtl8710bx-4mb-980k
  framework:
    version: latest

logger:

packages:
  # Uncomment exactly one line depending on your device type
  inv:   !include common/solis-modbus-inv.yaml
  #esinv: !include common/solis-modbus-esinv.yaml
  #epm:   !include common/solis-modbus-epm.yaml

  # Uncomment for optional INV inverter addon: Single Phase ACR10R energy meter
  #inv-acr10r-1ph: !include common/solis-modbus-inv-addon-acr10r-1ph.yaml

  # Uncomment for optional ESINV inverter addon: timers for battery charge and discharge periods
  #esinv-timer: !include common/solis-modbus-esinv-addon-timer.yaml

time:
  - platform: homeassistant
    id: network_time

api:
  encryption:
    key: !secret api_encryption_key

ota:
  - platform: esphome
    password: !secret ota_password

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  ap:
    ssid: "Ginlong Solis Inverter"
    password: !secret wifi_ap_password

captive_portal:

interval:
  - interval: 1s
    then:
      if:
        condition:
          wifi.connected:
        then:
          if:
            condition:
              switch.is_off: led_green_net
            then:
              - switch.turn_on: led_green_net
        else:
          if:
            condition:
              switch.is_on: led_green_net
            then:
              - switch.turn_off: led_green_net

script:
  - mode: restart
    id: modbus_activity
    then:
      - switch.turn_on: led_orange_com
      - if:
          condition:
            api.connected:
          then:
            - switch.turn_on: wdt_reset

uart:
  id: modbus_uart
  tx_pin: PA23
  rx_pin: PA18
  baud_rate: 9600

modbus:
  flow_control_pin: PA19
  send_wait_time: 300ms
  id: modbus_hw

modbus_controller:
  - id: modbus_master
    address: 0x1
    modbus_id: modbus_hw
    setup_priority: -10

button:
  - platform: factory_reset
    name: Restart with factory default settings
    id: reset_all
  - platform: restart
    name: Reboot device

switch:
  - platform: gpio
    restore_mode: ALWAYS_OFF
    pin: PA12
    id: led_orange_com
    on_turn_on:
    - delay: 500ms
    - switch.turn_off: led_orange_com
  - platform: gpio
    restore_mode: ALWAYS_OFF
    pin: PA05
    id: led_green_net
  - platform: gpio
    pin: PA00
    id: wdt_reset
    on_turn_on:
    - delay: 100ms
    - switch.turn_off: wdt_reset

binary_sensor:
  - platform: gpio
    pin:
      number: PA08
      inverted: true
    id: button_reset
    on_click:
      min_length: 5s
      max_length: 300s
      then:
        - button.press: reset_all

sensor:
  - platform: wifi_signal
    name: "WiFi Signal"
```

### Anything in the logs that might be useful for us?

```txt
Dependency Graph
|-- AsyncTCP @ 3.4.5
|-- ESPAsyncWebServer @ 3.9.6
|-- DNSServer @ 1.1.0
|-- noise-c @ 0.1.11
Compiling .pioenvs/solis-emw3080/src/main.cpp.o
Compiling .pioenvs/solis-emw3080/lib340/ESPAsyncWebServer/AsyncWebServerRequest.cpp.o
Compiling .pioenvs/solis-emw3080/lib340/ESPAsyncWebServer/AsyncWebSocket.cpp.o
Compiling .pioenvs/solis-emw3080/lib340/ESPAsyncWebServer/BackPort_SHA1Builder.cpp.o
Compiling .pioenvs/solis-emw3080/lib340/ESPAsyncWebServer/ChunkPrint.cpp.o
In file included from /data/cache/platformio/platforms/libretiny/cores/realtek-amb/base/fixups/basic_types.h:17,
                 from /data/cache/platformio/platforms/libretiny/cores/realtek-amb/base/fixups/platform/../basic_types.h:17,
                 from /data/cache/platformio/platforms/libretiny/cores/realtek-amb/base/fixups/platform/../platform_stdlib.h:18,
                 from /data/cache/platformio/platforms/libretiny/cores/realtek-amb/base/fixups/platform/platform_stdlib.h:3,
                 from /data/cache/platformio/packages/framework-realtek-amb1/component/common/api/network/include/lwipopts.h:27,
                 from /data/cache/platformio/platforms/libretiny/cores/common/base/config/lwipopts.h:12,
                 from /data/cache/platformio/platforms/libretiny/cores/realtek-amb/base/config/lwipopts.h:5,
                 from .pioenvs/solis-emw3080/include/lwipopts.h:1,
                 from /data/cache/platformio/packages/library-lwip/src/include/lwip/opt.h:51,
                 from /data/cache/platformio/packages/library-lwip/src/include/lwip/sockets.h:42,
                 from src/esphome/components/socket/headers.h:133,
                 from src/esphome/components/socket/socket.h:7,
                 from src/esphome/components/api/api_frame_helper.h:12,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:4,
                 from src/main.cpp:3:
/data/cache/platformio/packages/framework-realtek-amb1/component/soc/realtek/8711b/fwlib/include/basic_types.h:31:17: error: expected identifier before numeric constant
   31 | #define SUCCESS 0
      |                 ^
src/esphome/components/uart/uart_component.h:34:3: note: in expansion of macro 'SUCCESS'
   34 |   SUCCESS,          ///< Confirmed: all bytes left the TX FIFO.
      |   ^~~~~~~
/data/cache/platformio/packages/framework-realtek-amb1/component/soc/realtek/8711b/fwlib/include/basic_types.h:31:17: error: expected '}' before numeric constant
   31 | #define SUCCESS 0
      |                 ^
src/esphome/components/uart/uart_component.h:34:3: note: in expansion of macro 'SUCCESS'
   34 |   SUCCESS,          ///< Confirmed: all bytes left the TX FIFO.
      |   ^~~~~~~
In file included from src/esphome/components/uart/uart.h:7,
                 from src/esphome/components/modbus/modbus.h:4,
                 from src/esphome.h:48,
                 from src/main.cpp:3:
src/esphome/components/uart/uart_component.h:33:24: note: to match this '{'
   33 | enum class FlushResult {
      |                        ^
In file included from /data/cache/platformio/platforms/libretiny/cores/realtek-amb/base/fixups/basic_types.h:17,
                 from /data/cache/platformio/platforms/libretiny/cores/realtek-amb/base/fixups/platform/../basic_types.h:17,
                 from /data/cache/platformio/platforms/libretiny/cores/realtek-amb/base/fixups/platform/../platform_stdlib.h:18,
                 from /data/cache/platformio/platforms/libretiny/cores/realtek-amb/base/fixups/platform/platform_stdlib.h:3,
                 from /data/cache/platformio/packages/framework-realtek-amb1/component/common/api/network/include/lwipopts.h:27,
                 from /data/cache/platformio/platforms/libretiny/cores/common/base/config/lwipopts.h:12,
                 from /data/cache/platformio/platforms/libretiny/cores/realtek-amb/base/config/lwipopts.h:5,
                 from .pioenvs/solis-emw3080/include/lwipopts.h:1,
                 from /data/cache/platformio/packages/library-lwip/src/include/lwip/opt.h:51,
                 from /data/cache/platformio/packages/library-lwip/src/include/lwip/sockets.h:42,
                 from src/esphome/components/socket/headers.h:133,
                 from src/esphome/components/socket/socket.h:7,
                 from src/esphome/components/api/api_frame_helper.h:12,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:4,
                 from src/main.cpp:3:
/data/cache/platformio/packages/framework-realtek-amb1/component/soc/realtek/8711b/fwlib/include/basic_types.h:31:17: error: expected unqualified-id before numeric constant
   31 | #define SUCCESS 0
      |                 ^
src/esphome/components/uart/uart_component.h:34:3: note: in expansion of macro 'SUCCESS'
   34 |   SUCCESS,          ///< Confirmed: all bytes left the TX FIFO.
      |   ^~~~~~~
In file included from src/esphome/components/uart/uart.h:7,
                 from src/esphome/components/modbus/modbus.h:4,
                 from src/esphome.h:48,
                 from src/main.cpp:3:
src/esphome/components/uart/uart_component.h:86:11: error: 'FlushResult' does not name a type
   86 |   virtual FlushResult flush() = 0;
      |           ^~~~~~~~~~~
src/esphome/components/uart/uart_component.h:95:19: error: 'InternalGPIOPin' has not been declared
   95 |   void set_tx_pin(InternalGPIOPin *tx_pin) { this->tx_pin_ = tx_pin; }
      |                   ^~~~~~~~~~~~~~~
src/esphome/components/uart/uart_component.h:99:19: error: 'InternalGPIOPin' has not been declared
   99 |   void set_rx_pin(InternalGPIOPin *rx_pin) { this->rx_pin_ = rx_pin; }
      |                   ^~~~~~~~~~~~~~~
src/esphome/components/uart/uart_component.h:103:29: error: 'InternalGPIOPin' has not been declared
  103 |   void set_flow_control_pin(InternalGPIOPin *flow_control_pin) { this->flow_control_pin_ = flow_control_pin; }
      |                             ^~~~~~~~~~~~~~~
src/esphome/components/uart/uart_component.h:151:19: error: 'UARTParityOptions' has not been declared
  151 |   void set_parity(UARTParityOptions parity) { this->parity_ = parity; }
      |                   ^~~~~~~~~~~~~~~~~
src/esphome/components/uart/uart_component.h:155:3: error: 'UARTParityOptions' does not name a type
  155 |   UARTParityOptions get_parity() const { return this->parity_; }
      |   ^~~~~~~~~~~~~~~~~
src/esphome/components/uart/uart_component.h:202:3: error: 'InternalGPIOPin' does not name a type
  202 |   InternalGPIOPin *tx_pin_{};
      |   ^~~~~~~~~~~~~~~
src/esphome/components/uart/uart_component.h:203:3: error: 'InternalGPIOPin' does not name a type
  203 |   InternalGPIOPin *rx_pin_{};
      |   ^~~~~~~~~~~~~~~
src/esphome/components/uart/uart_component.h:204:3: error: 'InternalGPIOPin' does not name a type
  204 |   InternalGPIOPin *flow_control_pin_{};
      |   ^~~~~~~~~~~~~~~
src/esphome/components/uart/uart_component.h:213:3: error: 'UARTParityOptions' does not name a type
  213 |   UARTParityOptions parity_{UART_CONFIG_PARITY_NONE};
      |   ^~~~~~~~~~~~~~~~~
src/esphome/components/uart/uart_component.h: In member function 'void UARTComponent::set_tx_pin(int*)':
src/esphome/components/uart/uart_component.h:95:52: error: 'class UARTComponent' has no member named 'tx_pin_'
   95 |   void set_tx_pin(InternalGPIOPin *tx_pin) { this->tx_pin_ = tx_pin; }
      |                                                    ^~~~~~~
src/esphome/components/uart/uart_component.h: In member function 'void UARTComponent::set_rx_pin(int*)':
src/esphome/components/uart/uart_component.h:99:52: error: 'class UARTComponent' has no member named 'rx_pin_'
   99 |   void set_rx_pin(InternalGPIOPin *rx_pin) { this->rx_pin_ = rx_pin; }
      |                                                    ^~~~~~~
src/esphome/components/uart/uart_component.h: In member function 'void UARTComponent::set_flow_control_pin(int*)':
src/esphome/components/uart/uart_component.h:103:72: error: 'class UARTComponent' has no member named 'flow_control_pin_'; did you mean 'set_flow_control_pin'?
  103 |   void set_flow_control_pin(InternalGPIOPin *flow_control_pin) { this->flow_control_pin_ = flow_control_pin; }
      |                                                                        ^~~~~~~~~~~~~~~~~
      |                                                                        set_flow_control_pin
src/esphome/components/uart/uart_component.h: In member function 'void UARTComponent::set_parity(int)':
src/esphome/components/uart/uart_component.h:151:53: error: 'class UARTComponent' has no member named 'parity_'
  151 |   void set_parity(UARTParityOptions parity) { this->parity_ = parity; }
      |                                                     ^~~~~~~
src/esphome/components/uart/uart_component.h: At global scope:
src/esphome/components/uart/uart_component.h:219:1: error: expected declaration before '}' token
  219 | }  // namespace esphome::uart
      | ^
In file included from src/esphome/components/modbus/modbus.h:4,
                 from src/esphome.h:48,
                 from src/main.cpp:3:
src/esphome/components/uart/uart.h: In member function 'esphome::uart::FlushResult esphome::uart::UARTDevice::flush()':
src/esphome/components/uart/uart.h:48:47: error: 'class UARTComponent' has no member named 'flush'
   48 |   FlushResult flush() { return this->parent_->flush(); }
      |                                               ^~~~~
Compiling .pioenvs/solis-emw3080/lib340/ESPAsyncWebServer/Middleware.cpp.o
In file included from src/esphome.h:87,
                 from src/main.cpp:3:
src/esphome/components/uart/uart_component_libretiny.h: At global scope:
src/esphome/components/uart/uart_component_libretiny.h:25:15: error: 'esphome::uart::FlushResult esphome::uart::LibreTinyUARTComponent::flush()' marked 'override', but does not override
   25 |   FlushResult flush() override;
      |               ^~~~~
src/main.cpp: In function 'void setup()':
src/main.cpp:440:27: error: cannot convert 'esphome::libretiny::ArduinoInternalGPIOPin*' to 'int*'
  440 |   modbus_uart->set_tx_pin(libretiny_arduinointernalgpiopin_id);
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                           |
      |                           esphome::libretiny::ArduinoInternalGPIOPin*
In file included from src/esphome/components/uart/uart.h:7,
                 from src/esphome/components/modbus/modbus.h:4,
                 from src/esphome.h:48,
                 from src/main.cpp:3:
src/esphome/components/uart/uart_component.h:95:36: note:   initializing argument 1 of 'void UARTComponent::set_tx_pin(int*)'
   95 |   void set_tx_pin(InternalGPIOPin *tx_pin) { this->tx_pin_ = tx_pin; }
      |                   ~~~~~~~~~~~~~~~~~^~~~~~
src/main.cpp:444:27: error: cannot convert 'esphome::libretiny::ArduinoInternalGPIOPin*' to 'int*'
  444 |   modbus_uart->set_rx_pin(libretiny_arduinointernalgpiopin_id_2);
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                           |
      |                           esphome::libretiny::ArduinoInternalGPIOPin*
In file included from src/esphome/components/uart/uart.h:7,
                 from src/esphome/components/modbus/modbus.h:4,
                 from src/esphome.h:48,
                 from src/main.cpp:3:
src/esphome/components/uart/uart_component.h:99:36: note:   initializing argument 1 of 'void UARTComponent::set_rx_pin(int*)'
   99 |   void set_rx_pin(InternalGPIOPin *rx_pin) { this->rx_pin_ = rx_pin; }
      |                   ~~~~~~~~~~~~~~~~~^~~~~~
Compiling .pioenvs/solis-emw3080/lib340/ESPAsyncWebServer/WebAuthentication.cpp.o
*** [.pioenvs/solis-emw3080/src/main.cpp.o] Error 1
========================= [FAILED] Took 63.54 seconds =========================
```

### Additional information

_No response_

--- a/esphome/components/uart/uart_component.h
+++ b/esphome/components/uart/uart_component.h
@@ -30,12 +30,17 @@ enum UARTDirection {
 const LogString *parity_to_str(UARTParityOptions parity);
 
 /// Result of a flush() call.
+// Some vendor SDKs (e.g., Realtek) define SUCCESS as a macro.
+// Save and restore around the enum to avoid collisions with our scoped enum value.
+#pragma push_macro("SUCCESS")
+#undef SUCCESS
 enum class FlushResult {
   SUCCESS,          ///< Confirmed: all bytes left the TX FIFO.
   TIMEOUT,          ///< Confirmed: timed out before TX completed.
   FAILED,           ///< Confirmed: driver or hardware error.
   ASSUMED_SUCCESS,  ///< Platform cannot report result; success is assumed.
 };
+#pragma pop_macro("SUCCESS")
 
 class UARTComponent {
  public:

--- a/tests/components/uart/test.rtl87xx-ard.yaml
+++ b/tests/components/uart/test.rtl87xx-ard.yaml
@@ -1,0 +1,14 @@
+uart:
+  - id: uart_id
+    tx_pin: PA23
+    rx_pin: PA18
+    baud_rate: 9600
+    data_bits: 8
+    parity: NONE
+    stop_bits: 1
+
+switch:
+  - platform: uart
+    name: "UART Switch"
+    uart_id: uart_id
+    data: [0x01, 0x02, 0x03]

